subsys/net/lib/tftp/tftp_client.c

```diff
@@ -5,292 +5,247 @@
-	}
+	} while (tx_count <= TFTP_REQ_RETX);
 
-	LOG_DBG("Are we re-transmitting: NO");
-	return ERROR_OPCODE;
+	return ret;
 }
 
-/* Name: tftp_get
- * Description: This function gets "file" from the remote server.
- */
-int tftp_get(struct sockaddr *server, struct tftpc *client,
+int tftp_get(struct sockaddr *server_addr, struct tftpc *client,
 	     const char *remote_file, const char *mode)
 {
-
-	int32_t   stat     = TFTPC_UNKNOWN_FAILURE;
-	uint8_t    retx_cnt = 0;
-	int32_t   sock;
-
-	/* Re-init the global "block number" variable. */
-	tftpc_block_no = 1;
-	tftpc_index    = 0;
-
-	/* Create Socket for TFTP (IPv4 / UDP) */
-	sock = socket(server->sa_family, SOCK_DGRAM, IPPROTO_UDP);
+	int sock;
+	uint32_t tftpc_block_no = 1;
+	uint32_t tftpc_index = 0;
+	int tx_count = 0;
+	struct tftphdr_ack ackhdr = {htons(ACK_OPCODE), htons(1)};
+	int rcv_size;
+	int ret;
+
+	sock = socket(server_addr->sa_family, SOCK_DGRAM, IPPROTO_UDP);
 	if (sock < 0) {
 		LOG_ERR("Failed to create UDP socket (IPv4): %d", errno);
 		return -errno;
 	}
 
-	/* Connect with the address.  */
-	stat = connect(sock, server, sizeof(struct sockaddr_in6));
-	if (stat < 0) {
-		LOG_ERR("Cannot connect to UDP remote (IPv4): %d", errno);
-		return -errno;
-	}
-
 	/* Obtain Global Lock before accessing critical resources. */
 	k_mutex_lock(&tftpc_lock, K_FOREVER);
 
 	/* Send out the request to the TFTP Server. */
-	stat = tftp_send_request(sock, READ_REQUEST, remote_file, mode);
-	if (stat == ERROR_OPCODE) {
-		LOG_ERR("Server responded with error.");
-		return TFTPC_REMOTE_ERROR;
-	}
+	ret = tftp_send_request(sock, server_addr, READ_REQUEST, remote_file, mode);
+	rcv_size = ret;
 
-process_resp:
+	while (rcv_size >= TFTP_HEADER_SIZE && rcv_size <= TFTPC_MAX_BUF_SIZE) {
+		/* Process server response. */
 
-	/* Process server response. */
-	stat = tftpc_process_resp(sock, client);
-	if (stat == TFTPC_BUFFER_OVERFLOW ||
-	    stat == TFTPC_SUCCESS) {
-		goto req_done;
-	}
+		uint16_t opcode = sys_get_be16(tftpc_buffer);
+		uint16_t block_no = sys_get_be16(tftpc_buffer + 2);
 
-tftpc_recv:
+		LOG_DBG("Received data: opcode %u, block no %u, size %d",
+			opcode, block_no, rcv_size);
 
-	/* More data is available - Read (or we read a duplicate). */
-	stat = tftpc_recv_data(sock);
-	if (stat <= 0) {
-		/* No response from server. */
-		LOG_DBG("Server response timeout.");
+		if (opcode != DATA_OPCODE) {
+			LOG_ERR("Server responded with invalid opcode.");
+			ret = TFTPC_REMOTE_ERROR;
+			break;
+		}
 
-		/* Retries exhausted ? */
-		if (++retx_cnt >= TFTP_REQ_RETX) {
-			LOG_ERR("No more retransmits available. Exiting");
-			return TFTPC_RETRIES_EXHAUSTED;
+		if (block_no == tftpc_block_no) {
+			uint32_t data_size = rcv_size - TFTP_HEADER_SIZE;
+
+			tftpc_block_no++;
+			ackhdr.block = htons(block_no);
+			tx_count = 0;
+
+			/* Only copy block if user buffer has enough space */
+			if (data_size > (client->user_buf_size - tftpc_index)) {
+				send_err(sock, 0x3, NULL);
```

Comment:
Could you add a symbol for this magic number?

Reply:
I've added respective defines to `subsys/net/lib/tftp/tftp_client.h`. The comments describing errors are copied verbatim from [RFC 1350](https://datatracker.ietf.org/doc/html/rfc1350). I expect that falls under fair use. To be on the safe side I can remove them or try to rephrase.